File: algorithms/1_easy/S/02996_smallest_missing_integer_greater_than_sequential_prefix_sum/solution.c
#define MAX_NUM 50
#define MIN_NUM 1
#define MAX_LEN 50

int missingInteger(const int* nums, const int len) {
    if (nums == NULL || len > MAX_LEN || len <= 0) { goto input_err; }

    bool seen[MAX_NUM + 2] = { false };
    for (int i = 0; i < len; i += 1) {
        if (nums[i] < MIN_NUM || nums[i] > MAX_NUM) {
            goto input_err;
        }
        seen[nums[i]] = true;
    }

    int prefix = nums[0];
    int i = 1;
    while (i < len && nums[i] == nums[i - 1] + 1) {
        prefix += nums[i];
        i += 1;
    }

    if (prefix > MAX_NUM) {
        return prefix;
    }

    for (int i = prefix; i < MAX_NUM + 2; i += 1) {
        if (!seen[i]) {
            return i;
        }
    }

    fprintf(stderr, "unreachable!()");
    exit(1);

input_err:
    fprintf(stderr, "invalid input");
    return -1;
}

File: algorithms/1_easy/S/02996_smallest_missing_integer_greater_than_sequential_prefix_sum/solution.rs
const MAX_NUM: i32 = 50;
const MIN_NUM: i32 = 1;

impl Solution {
    pub fn missing_integer(nums: Vec<i32>) -> i32 {
        assert!(!nums.is_empty() && nums[0] >= MIN_NUM && nums[0] <= MAX_NUM);
        let len = nums.len();

        let mut seen = [false; MAX_NUM as usize + 2];
        for &num in &nums {
            assert!(num >= MIN_NUM && num <= MAX_NUM);
            seen[num as usize] = true;
        }

        let mut prefix = nums[0];
        let mut i = 1_usize;
        while i < len && nums[i] == nums[i - 1] + 1 {
            prefix += nums[i];
            i += 1;
        }

        if prefix > MAX_NUM {
            return prefix;
        }

        for i in prefix..MAX_NUM + 2 {
            if !seen[i as usize] {
                return i;
            }
        }

        unreachable!()
    }
}
